py/dml/types.py

```diff
@@ -279,7 +328,23 @@ def print_declaration(self, var, init = None, unused = False):
     def describe(self):
         raise Exception("%s.describe not implemented" % self.__class__.__name__)
 
+    def hashed(self):
+        '''Hash the DML type in a way compatible with eq. I.e.
+           a.eq(b) implies a.hashed() == b.hashed()'''
+        assert type(self).eq is DMLType.eq, \
+               '.eq() overridden without overriding .hashed()'
+        return hash((type(self), self.const))
+
     def key(self):
+        '''A string key uniquely identifying the DML type, and is suitable for
+        use in checkpointing. Will raise DMLUnkeyableType if such a key can't
+        be produced.
+
+        If neither a, b is unkeyable, then:
+        * a.eq(b) must imply a.key() == b.key()
+        * a.key() == b.key() implies a.eq(b) with overwhelming likelihood
+          (though not necessarily guaranteed; see IntegerType.key())
+        '''
```

Comment:
The changes to `key()` in this PR technically causes backwards checkpoint incompatibility... were it not for the fact that it's `key()` currently only used for `hook.send_now()` callbacks of afters. Making it completely irrelevant in practice.